### PR TITLE
fhir-persistence-schema CLI drop schema gives an error #2758

### PR DESCRIPTION
**Pull request by prb112**
- Add missing tags to whole system tables to enforce a FK drop prior to
table drop
- Add tags to the OAuthSchemaGenerator and JavaBatchGenerator (drop
wasn't working without the tags)
- Tags are unique for the Generator

No schema change this is in the runtime of the CLI only.

Signed-off-by: Paul Bastide <pbastide@us.ibm.com>

### CODE REVIEW

File: fhir-persistence-schema/src/main/java/com/ibm/fhir/schema/control/JavaBatchSchemaGenerator.java [hunk 1]
```diff
@@ -142,23 +142,26 @@ VALUE VARCHAR(255),
...
                 .addVarcharColumn(VALUE, 4096, NULL) // VALUE VARCHAR(255) (updated to 4096)
                 .addBigIntColumn(FK_JOBEXECID, NOT_NULL) // FK_JOBEXECID BIGINT
                 .addForeignKeyConstraintAltTarget("JBPRMETERFKJBXECID", schemaName, JOBEXECUTION_TABLE, JOBEXECID, FK_JOBEXECID)
                 .addIndex("JP_FKJOBEXECID_IX", FK_JOBEXECID)
                 .addPrivileges(generateGroupPrivilege())
                 .build(model);
-        model.addTable(parameter);
-        model.addObject(parameter);
+
+        table.addTag(FhirSchemaGenerator.SCHEMA_GROUP_TAG, FhirSchemaGenerator.FHIRDATA_GROUP);
```
Comment (lmsurpre) [inline fhir-persistence-schema/src/main/java/com/ibm/fhir/schema/control/JavaBatchSchemaGenerator.java:154]: should we use a different group id for the jbatch schema objects?

File: fhir-persistence-schema/src/main/java/com/ibm/fhir/schema/control/OAuthSchemaGenerator.java [hunk 2]
```diff
@@ -102,6 +102,7 @@ protected void addCacheTable(PhysicalDataModel model) {
                 .addUniqueIndex(OAUTH20CACHE + "_" + EXPIRES, EXPIRES) // ASC is the default
                 .addPrivileges(generateGroupPrivilege())
                 .build(model);
+        cache.addTag(FhirSchemaGenerator.SCHEMA_GROUP_TAG, FhirSchemaGenerator.FHIRDATA_GROUP);
```
Comment (lmsurpre) [inline fhir-persistence-schema/src/main/java/com/ibm/fhir/schema/control/OAuthSchemaGenerator.java:105]: should we use a different group id for the OAuth schema objects?

Comment (prb112) [reply fhir-persistence-schema/src/main/java/com/ibm/fhir/schema/control/OAuthSchemaGenerator.java:105]: I think that implied much more work were we split out the Physical Data model.  If not true, is that OK to change... @punktilious ?

Comment (lmsurpre) [reply fhir-persistence-schema/src/main/java/com/ibm/fhir/schema/control/OAuthSchemaGenerator.java:105]: yeah, i just had a quick look in Main and we have logic like this:
```
if (dropFhirSchema || dropOauthSchema || dropJavaBatchSchema) {
            logger.info("Dropping FK constraints in the data schema: " + this.schema.getSchemaName());
            dropForeignKeyConstraints(pdm, FhirSchemaGenerator.SCHEMA_GROUP_TAG, FhirSchemaGenerator.FHIRDATA_GROUP);
        }

...

if (dropFhirSchema || dropOauthSchema || dropJavaBatchSchema) {
                        // Just drop the objects associated with the FHIRDATA schema group
                        pdm.drop(adapter, FhirSchemaGenerator.SCHEMA_GROUP_TAG, FhirSchemaGenerator.FHIRDATA_GROUP);
                    }
```

I guess I don't really understand what the "group" is supposed to signify here.

Comment (prb112) [reply fhir-persistence-schema/src/main/java/com/ibm/fhir/schema/control/OAuthSchemaGenerator.java:105]: I tried it out. It'll fail miserably. it has to be part of separating the physical model.

File: fhir-persistence-schema/src/main/java/com/ibm/fhir/schema/control/JavaBatchSchemaGenerator.java (hunk 1 above)
Comment (prb112) [reply fhir-persistence-schema/src/main/java/com/ibm/fhir/schema/control/JavaBatchSchemaGenerator.java:154]: see the following comment about PhysicalDataModel

File: fhir-persistence-schema/src/main/java/com/ibm/fhir/schema/control/OAuthSchemaGenerator.java (hunk 2 above)
Comment (prb112) [reply fhir-persistence-schema/src/main/java/com/ibm/fhir/schema/control/OAuthSchemaGenerator.java:105]: it's part of the PhysicalDataModel

Comment (lmsurpre) [reply fhir-persistence-schema/src/main/java/com/ibm/fhir/schema/control/OAuthSchemaGenerator.java:105]: Do we know anywhere else that the SCHEMA_GROUP_TAG is used?

Comment (prb112) [reply fhir-persistence-schema/src/main/java/com/ibm/fhir/schema/control/OAuthSchemaGenerator.java:105]: got it to work.... changed the tag names

Comment (lmsurpre) [reply fhir-persistence-schema/src/main/java/com/ibm/fhir/schema/control/OAuthSchemaGenerator.java]: yes.  this looks exactly how i was expecting it.  marking LGTM but lets wait for robin's review to be sure we're not missing something!

**Review by lmsurpre — APPROVED** [review]
LGTM

File: fhir-persistence-schema/src/main/java/com/ibm/fhir/schema/control/FhirSchemaGenerator.java [hunk 3]
```diff
@@ -717,6 +717,9 @@ public Table addLogicalResourceTags(PhysicalDataModel pdm) {
                 })
                 .build(pdm);
 
+        tbl.addTag(SCHEMA_GROUP_TAG, FHIRDATA_GROUP);
+        this.procedureDependencies.add(tbl);
+
         pdm.addTable(tbl);
         pdm.addObject(tbl);
```
Comment (punktilious) [inline fhir-persistence-schema/src/main/java/com/ibm/fhir/schema/control/FhirSchemaGenerator.java:724]: Do you think it's worth adding a check to make sure there's at least one tag on the object? A better change (although a lot more work) would be to make tagging part of the builder and then we could make it mandatory.

**Review by punktilious — COMMENTED** [review]
Should we extend the schema migration test to include a drop phase...then we can check that all the objects have been dropped.

File: fhir-persistence-schema/src/main/java/com/ibm/fhir/schema/control/FhirSchemaGenerator.java (hunk 3 above)
Comment (prb112) [reply fhir-persistence-schema/src/main/java/com/ibm/fhir/schema/control/FhirSchemaGenerator.java:724]: I've added the checks in place to tests, and modified where the tags are missing. 
The build will fail without it.

**Pull request merged**